### PR TITLE
Move Ternary from std.experimental.allocator.common to std.typecons

### PR DESCRIPTION
**Pull request by JackStouffer**
This needs to be moved before std.experimental.allocator is moved anyway, better to do it sooner rather than later.

ping @andralex

### CODE REVIEW

**Comment by ZombineDev** [comment]
Just FYI, there is a `std.experimental.typecons` module. I'm not sure if Ternary should go there first or not.

File: std/typecons.d [hunk 1]
```diff
@@ -7143,3 +7143,150 @@ unittest
         ubyte, ubyte, T3, T3,
     );
 }
+
+/**
+Ternary type with three thruth values.
+*/
```
Comment (wilzbach) [inline std/typecons.d]: I know it's redundant, but it probably doesn't hurt to state those three values here

File: std/typecons.d [hunk 2]
```diff
@@ -7143,3 +7143,150 @@ unittest
...
+    Ternary opBinary(string s)(Ternary rhs) if (s == "&")
+    {
+        return make(26_144 >> value + rhs.value & 6);
+    }
+
+    /// ditto
+    Ternary opBinary(string s)(Ternary rhs) if (s == "^")
+    {
+        return make(26_504 >> value + rhs.value & 6);
+    }
```
Comment (wilzbach) [inline std/typecons.d]: How about overloading the increment and decrement operators?

Comment (JackStouffer) [inline std/typecons.d]: Enhancements can be added later.

**Comment by JackStouffer** [comment]
Beefed up the documentation.

**Comment by wilzbach** [comment]
> Just FYI, there is a std.experimental.typecons module. I'm not sure if Ternary should go there first or not.

I guess that's about Andrei to decide, but the current policy for adding new names seems to be "direct add". Speaking of which - can someone please label this? Thanks!

**Comment by JackStouffer** [comment]
Ping.

**Comment by wilzbach** [comment]
The only way to reach Andrei atm is via personal mail.

File: std/typecons.d [hunk 3]
```diff
@@ -7143,3 +7143,167 @@ unittest
         ubyte, ubyte, T3, T3,
     );
 }
+
+/**
+Ternary type with three truth values: `true`, `false`, and `unknown`.
```
Comment (andralex) [inline std/typecons.d]: This is the place where you should specify they have the names `yes`, `no`, and `unknown` respectively.

File: std/typecons.d [hunk 4]
```diff
@@ -7143,3 +7143,167 @@ unittest
...
+
+See_Also:
+    $(WEB https://en.wikipedia.org/wiki/Three-valued_logic,
+        Three Valued Logic on Wikipedia)
+*/
+struct Ternary
+{
+    @safe @nogc nothrow pure:
+
+    package ubyte value = 6;
```
Comment (andralex) [inline std/typecons.d]: why is `package` needed?

File: std/typecons.d [hunk 5]
```diff
@@ -7143,3 +7143,167 @@ unittest
...
+    package static Ternary make(ubyte b)
```
Comment (andralex) [inline std/typecons.d]: why `package`?

File: std/typecons.d [hunk 6]
```diff
@@ -7143,3 +7143,167 @@ unittest
...
+    */
+    Ternary opUnary(string s)() if (s == "~")
+    {
+        return make(386 >> value & 6);
+    }
+
+    /// ditto
+    Ternary opBinary(string s)(Ternary rhs) if (s == "|")
+    {
+        return make(25_512 >> value + rhs.value & 6);
```
Comment (andralex) [inline std/typecons.d]: some parens would be nice - I at least am not sure what's going on :o)

File: std/typecons.d [hunk 7]
```diff
@@ -7143,3 +7143,167 @@ unittest
...
+
+    /// ditto
+    Ternary opBinary(string s)(Ternary rhs) if (s == "^")
+    {
+        return make(26_504 >> value + rhs.value & 6);
+    }
+}
+
+///
+unittest
```
Comment (andralex) [inline std/typecons.d]: please add pure @nogc etc

**Comment by andralex** [comment]
Very nicely executed, thanks. I approve this. @JackStouffer, after you take a look at a few nits you may pull this yourself.

**Comment by andralex** [comment]
@JackStouffer actually let's have someone else pull this just to follow protocol. Thanks!

**Comment by JackStouffer** [comment]
@andralex I don't even have merge rights, so no problem ;)

**Comment by andralex** [comment]
mmm okay I'll be watching this, let me know if you're done making changes (all are optional in fact)

**Comment by JackStouffer** [comment]
@andralex Fixed comments and added change log entry

**Comment by andralex** [comment]
nice, thanks!

**Comment by JackStouffer** [comment]
Thanks!

**Comment by wilzbach** [comment]
What happened to the policy that all additions need to go through auto-tester?

**Comment by JackStouffer** [comment]
Technically speaking it did, just not for the latest rebased commit.

**Pull request merged**